magma/circuit.py

```diff
@@ -587,6 +633,12 @@ def bind(cls, monitor, *args):
             content = "\n".join((f.read(), bind_str))
         cls.bind_modules[monitor.name] = content
 
+    def add_instance(cls, circuit, *args, **kwargs):
```

Comment:
This seems a little wonky -- if we want to just place the instance, we should do that directly in the pass rather than call a function on the circuit which is exposed to the world? Or this function `add_instance` should be something that can be expected to be used by outside users

Reply:
The issue is that then the user needs to do
```
        _DefinitionBlock.push(CurrentDefinition)
        inst = circuit(*args, **kwargs)
        _DefinitionBlock.pop()
```
Adding at as a method to the class allows us to easily abstract this logic (since the class has a reference to itself).  However, this also means we can't create the instance until the DefinitionBlock is populated.

I'd rather avoid the user having to push/pop the definition stack every time an instance is added.  I'm not sure if there's another way around this other than maybe adding a notion of a floating instance and placing it afterwards? The problem is that right now there is an implicit `place` into the top of the definition stack, which I'm not sure we can guarantee is empty when running a pass (in theory a user could use a pass while inside a definition).

Reply:
That makes sense...what about something like this:

```python
def open_definition(defn):
    return _PlacerBlock(defn._placer)
```

Then later, instead of 

```python
            wire_inst = definition.add_instance(
                Wire, T, name=name
            )
            self.wire_map[driver] = wire_inst
            if issubclass(T, Digital):
                wire_inst.I @= driver
            else:
                wire_inst.I @= driver.as_bits()
```

we could write

```python
            with open_definition(definition):
                wire_inst = Wire(T, name=name)
                self.wire_map[driver] = wire_inst
                if issubclass(T, Digital):
                    wire_inst.I @= driver
                else:
                    wire_inst.I @= driver.as_bits()
```

I like this pattern better because it makes the places where we edit circuits very explicit. I've thought about it a bit, and having circuit editing methods on the classes directly seems dangerous.